source/extensions/common/tap/admin.cc

```diff
@@ -0,0 +1,110 @@
+#include "extensions/common/tap/admin.h"
+
+#include "envoy/admin/v2alpha/tap.pb.h"
+#include "envoy/admin/v2alpha/tap.pb.validate.h"
+
+#include "common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Tap {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(tap_admin_handler);
+
+AdminHandlerSharedPtr AdminHandler::getSingleton(Server::Admin& admin,
+                                                 Singleton::Manager& singleton_manager,
+                                                 Event::Dispatcher& main_thread_dispatcher) {
+  return singleton_manager.getTyped<AdminHandler>(
+      SINGLETON_MANAGER_REGISTERED_NAME(tap_admin_handler), [&admin, &main_thread_dispatcher] {
+        return std::make_shared<AdminHandler>(admin, main_thread_dispatcher);
+      });
+}
+
+AdminHandler::AdminHandler(Server::Admin& admin, Event::Dispatcher& main_thread_dispatcher)
+    : admin_(admin), main_thread_dispatcher_(main_thread_dispatcher) {
+  bool rc =
+      admin_.addHandler("/tap", "tap filter control", MAKE_ADMIN_HANDLER(handler), true, true);
+  RELEASE_ASSERT(rc, "/tap admin endpoint is taken");
+}
+
+AdminHandler::~AdminHandler() {
+  bool rc = admin_.removeHandler("/tap");
+  ASSERT(rc);
+}
+
+Http::Code AdminHandler::handler(absl::string_view, Http::HeaderMap&, Buffer::Instance& response,
+                                 Server::AdminStream& admin_stream) {
+  if (attached_request_.has_value()) {
+    response.add("An attached /tap admin stream already exists. Detach it.");
```

Comment:
Is there a plan to support multiple streams? Seems like there's no need for this limit, even though it might make sense for the MVP.

Reply:
Yeah I just did this for simplicity. I will add a TODO/comment here.